### DISCUSSION TITLE
How-To: Start Workshop Example app and use the Java API in another project

### DISCUSSION THREAD

**Discussion by javac9** [openmdx/openmdx-example#5] [Q&A]
After fiddling with OpenCrx for some time, I am trying now to understand how OpenMdx itself works and how can I model myself a small example app.

I see the process as following:
- I should use one of the example projects as the basis for my small example app
- I can edit the model (for example, add an attribute to some existing class)
- I run the Gradle clean-assemble (and maybe also class?) tasks to get this new attribute generated
I tried it up to this point and I can see my dummy attribute generated.

- Now I could try to start the app and for example check if I can access the Rest API. How do I do that without an installer and possibly directly from Eclipse IDE?
- As a final step with my small example app, I might try to use Java API of my example app in another Java app. What JAR dependencies should I include in this Java app?

**Comment by javac9** [comment]
Hate to bump the Question but since I have some free time, it'd be great to try to make some progress.

**Comment by wfro64** [comment]
https://github.com/openmdx/openmdx-documentation/blob/master/Sdk/Workshop.md describes the development roundtrip. In addition, the section 'Deploy as REST service' has been added which explains how to deploy an openMDX application as REST service. For deploying the application as a Java app see https://github.com/openmdx/openmdx-example/blob/master/workshop/src/test/java/org/openmdx/test/example/workshop1/TestWorkshop_1.java. This runs the workshop app as JUnit.

**Comment by javac9** [comment]
The section https://github.com/openmdx/openmdx-documentation/blob/master/Sdk/Workshop.md#deploy-as-rest-service states `Copy openmdx-workshop-rest.war to apache-tomee-plus-8.0.3/webapps/` but I can't find it anywhere within the `Workshop` project. First place I looked was `../build/distributions` but there is only what is shown on the attached screenshot (also, note the file version - v2.17.8 instead of v2.17.9)
![openmdx jpg](https://user-images.githubusercontent.com/25393309/128692889-503ff670-5337-49ab-ba05-d2ff333cfe40.png)

**Comment by wfro64** [comment]
pull the latest changes from master https://github.com/openmdx/openmdx-example

**Reply by javac9** [reply]
Thanks. I just tried that - now I can see the REST war but in `../build/tmp` and as a folder (contains only MANIFEST file), not a WAR archive. The stuff from the above screenshot is same (version mismatch).

**Comment by wfro64** [comment]
The WAR in `./build/tmp` is not the correct one. The WAR is located in `./workshop/build/deployment-unit`. The content looks as follows:

![Screenshot_20210809_170148](https://user-images.githubusercontent.com/11387061/128728627-b742cdbe-812b-4952-a4ac-f58fe995d2d5.png)

**Reply by javac9** [reply]
I'll certainly check the said location as soon as can but what concerns me is that I used Windows search function to search within the whole workshop directory. I suspect if it were there, it would have appeared in the results. I guess it rather wasn't generated for some reason.

**Comment by wfro64** [comment]
I do 

```
git clone git@github.com:openmdx/openmdx-example.git openmdx-example
cd openmdx-example
gradle assemble
```

and it is there.

**Reply by javac9** [reply]
Ok, it was something on my side. I managed to deploy both `rest` and `portal` apps to TomEE.
In `rest` I get an error in UI: 
![Screenshot 2021-08-14 000902](https://user-images.githubusercontent.com/25393309/129422951-0d2ba8b0-2db5-4e9a-850c-bfaa3aa9fe8d.jpg)
and in console:
```
14-Aug-2021 00:06:32.473 WARNUNG [http-nio-8080-exec-8] org.openmdx.application.rest.http.servlet.RestServlet_2.handleException Sys|Resource Exception
org.openmdx.base.exception.ServiceException: DefaultDomain.MEDIA_ACCESS_FAILURE
    BasicException[0]
        Class = org.openmdx.base.dataprovider.layer.persistence.jdbc.Database_2
        Method = getPartialObject
        Line = 6827
        ExceptionClass = org.openmdx.base.exception.ServiceException
        ExceptionDomain = DefaultDomain
        ExceptionCode = MEDIA_ACCESS_FAILURE
        Timestamp = 2021-08-13T22:06:32.360Z
        Parameters:
            0:  xri = xri://@openmdx*org.openmdx.example.workshop1/provider/Workshop/segment/Standard
            1:  statement = SELECT  * FROM OOMEP1WS_SEGMENT v WHERE ((v.object_id = ?))
            2:  parameters = null
            3:  sqlErrorCode = -1305
            4:  sqlState = 08006
        StackTrace:
            at org.openmdx.base.dataprovider.layer.persistence.jdbc.Database_2.getPartialObject(Database_2.java:6827)
            at org.openmdx.base.dataprovider.layer.persistence.jdbc.Database_2.get(Database_2.java:6852)
            at org.openmdx.base.dataprovider.layer.persistence.jdbc.RestInteraction.get(RestInteraction.java:167)
            at org.openmdx.base.rest.spi.AbstractRestInteraction.handleQuery(AbstractRestInteraction.java:488)
            at org.openmdx.base.rest.spi.AbstractRestInter.....
```

**Reply by javac9** [reply]
Regarding the `portal` app, I was able to only login as `guest`. What do I need to do apart from adding a corresponding Tomcat user to be able to login as `admin-Standard`?

Note: I can see "volatile" in the app and none of the initial data (skills, etc.) is present (I can see the initial data through a Hsqldb manager). Also, during the TomEE startup I already get a connection error (attached screenshot).
![Screenshot 2021-08-14 004508](https://user-images.githubusercontent.com/25393309/129425189-3038f275-71d6-4c9b-87a5-178747389635.jpg)

**Comment by wfro64** [comment]
You have to start the HSQLDB database as described [here](https://github.com/openmdx/openmdx-documentation/blob/master/Sdk/Workshop.md):

```
cd etc/data/workshop/.
./startdb.sh
```
Otherwise TomEE is not able to connect and you get `MEDIA_ACCESS_FAILURE` and the WebApps won't work.

See Google how to add users to Tomcat "how to add users to tomcat users xml". Users must have role "Guest" which is required by [web.xml](https://github.com/openmdx/openmdx-example/blob/master/workshop/src/data/openmdx-workshop-portal/WEB-INF/web.xml)

**Reply by javac9** [reply]
I did start the HSQLDB that way and after that also the DB manager (`./startdbmanager.sh`) - that's why I said above I could see the initial data in the manager.

I'm using version 2.6.0 of Hsqldb and this is what I get in Hsqldb console right after I start TomEE:
```
[Server@52a86356]: Startup sequence initiated from main() method
[Server@52a86356]: Could not load properties from file
[Server@52a86356]: Using cli/default properties only
[Server@52a86356]: Initiating startup sequence...
[Server@52a86356]: Server socket opened successfully in 4 ms.
[Server@52a86356]: Database [index=0, id=0, db=file:workshop, alias=workshop] opened successfully in 176 ms.
[Server@52a86356]: Startup sequence completed in 181 ms.
[Server@52a86356]: 2021-08-14 07:55:13.575 HSQLDB server 2.6.0 is online on port 9002
[Server@52a86356]: To close normally, connect and execute SHUTDOWN SQL
[Server@52a86356]: From command line, use [Ctrl]+[C] to abort abruptly
org.hsqldb.HsqlException: HSQLDB server version is '2.6.0'  client version '2.1.0.0' is not compatible.
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.server.ServerConnection.init(Unknown Source)
        at org.hsqldb.server.ServerConnection.run(Unknown Source)
        at java.base/java.lang.Thread.run(Thread.java:834)
```
The file `./etc/data/workshop/workshop.properties` mentions version `2.6.0`:
```
#HSQL Database Engine 2.6.0
#Mon Aug 09 13:50:58 CEST 2021
tx_timestamp=291
modified=yes
version=2.6.0
```
Now, the question is: what is the correct version of Hsqldb to go with: `2.1.0.0` or `2.6.0`?

**Comment by wfro64** [comment]
In the [documentation](https://github.com/openmdx/openmdx-documentation/blob/master/Sdk/Workshop.md) it says:

_Upgrade the JDBC driver for HSQDB: copy hsqldb-2.5.0.jar to apache-tomee-plus-8.0.3/lib/ and remove hsqldb-2.3.2.jar. NOTE: the version of the driver must match with the version that was specified when launching the WORKSHOP database with the startup script located in ./etc/data/workshop/_

**Reply by javac9** [reply]
I made a stupid mistake when copying the jar to tomee lib folder - I took the `hsqldb.jar` from within my local hsqldb installation, simply renamed it to `hsqldb-2.5.0.jar` and copied it there. Now I downloaded the jar from here `https://repo1.maven.org/maven2/org/hsqldb/hsqldb/2.5.0/` and it worked. Thanks for your patience!!
